### ISSUE TITLE
moveFrom, moveInitialize: the compiler needs to be aware

### ISSUE THREAD

**Issue by vasslitvinov** [chapel-lang/chapel#20437]
These functions are also discussed in #20330 and #20328.

The charter of the `Memory.Initialization` module is "[to enable the user] to implement collections in a manner similar to those implemented by the Chapel standard modules (such as List or Map)."

I claim that at present the functions `moveFrom` and `moveInitialize` do not achieve this purpose.

#### Example 1: `LinkedList.pop_front()`

I expect pop_front() to pass to the caller the ownership of the element being popped:
```chpl
var oldFront = ... the front listNode ...;
var ret = moveFrom(oldFront.data);
delete oldFront;
return ret;
```

However, it does not -- and cannot do that. Because `delete oldFront` deinitializes `oldFront.data` among other things - because it does not know that `oldFront.data` has been "consumed". So if I called `moveFrom(oldFront.data)`, the deinitializer would deinitialize invalid memory contents.

#### Example 2: `chpl__hashtable.clearSlot()`

I expect the `out val` formal to take over the ownership of `tableEntry.val`:
```chpl
proc chpl__hashtable.clearSlot(ref tableEntry, out key: keyType, out val: valType) {
  moveInitialize(key, moveFrom(tableEntry.key));
  moveInitialize(val, moveFrom(tableEntry.val));
  ....
}
```

However, it does not do that. If it did, the compiler would not know that `key` and `value` would be initialized and would default-initialize them prior to the moveInitialize() calls. So moveInitialize() would clobber objects without deinitializing them properly. Instead, `key` and `value` get initialized using `=`:
```chpl
proc chpl__hashtable.clearSlot(ref tableEntry, out key: keyType, out val: valType) {
  key = moveFrom(tableEntry.key);
  val = moveFrom(tableEntry.val);
  ....
}
```

The problem with that is that the compiler inserts calls to chpl__coerceMove(), passing in the results of `moveFrom` and using the returned values to initialize `key`/`value`. chpl__coerceMove() is a Really Big Function and is completely unnecessary here.

In summary, this issue calls for compiler awareness of initialization and deinitialization performed by `moveFrom` and `moveInitialize`.

**Comment by mppf** [comment]
For Example 1, I expect that the data structure needs to use something lower level than `delete`. (Something more like `c_free`). I do not know how to make it work, otherwise.

For Example 2, I read it as suggesting that the compiler's split-init logic should consider "moveInitialize" as a way to initialize something.

**Comment by vasslitvinov** [comment]
These two functions request initialization and (`moveFrom`) deinitialization explicitly. However, it is not easy for the compiler to reason about it, especially when the affected memory locations (I call them "variables") are class/record fields or array elements.

To address that, what if we let the user annotate a variable (also array elements or a class/record field) with an attribute that tells the compiler "do NOT initialize / deinitialize, I will do it myself" ?

**Comment by mppf** [comment]
> To address that, what if we let the user annotate a variable [...] with an attribute that tells the compiler "do NOT initialize / deinitialize, I will do it myself" ?

I'm not opposed in principle but I suspect there is a better way to write it. Like, maybe we just need a function to "sink" the value and not deinit it.

> also array elements or a class/record field

I'm not convinced this is even possible, in general.

> These two functions request initialization and (`moveFrom`) deinitialization explicitly.

Huh? I don't think moveFrom requests any deinitialization.

**Comment by vasslitvinov** [comment]
> I don't think moveFrom requests any deinitialization.

Not "request", rather "perform".

Perhaps not "initialization" or "deinitialization" per se, rather switching from the uninitialized state to initialized (moveInitialize) and back (moveFrom).

> I'm not convinced this is even possible, in general.

My suggestion is for the user to associate the attribute(s):
* with a record variable, or
* with an array variable -- so the attribute(s) apply to all array elements, perhaps including those added or removed upon resizing the array, or
* with a field declaration -- so they apply to the field in all instances of the enclosing class/record.
